File: Benchmarks/CreateCSV.java

import java.io.*;
import java.util.Random;

public class CreateCSV {
    public static void main(String[] args) {
        String filename = "large_data_java.csv";
        int rows = 25_000_000;
        Random rand = new Random();

        // Έναρξη χρονομέτρησης
        long startTime = System.nanoTime();

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            writer.write("ID,Value1,Value2,Category\n");

            for (int i = 1; i <= rows; i++) {
                // Χρήση StringBuilder για καλύτερη απόδοση στην ένωση String
                writer.write(i + "," + rand.nextDouble() + "," + rand.nextInt() + ",ABCDE\n");

                if (i % 5_000_000 == 0) {
                    System.out.println("Πρόοδος: " + i + " γραμμές...");
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        // Λήξη χρονομέτρησης
        long endTime = System.nanoTime();
        double durationSeconds = (endTime - startTime) / 1_000_000_000.0;

        System.out.println("------------------------------------");
        System.out.println("Συνολικός χρόνος: " + String.format("%.2f", durationSeconds) + " δευτερόλεπτα");
        System.out.println("Ταχύτητα: " + String.format("%.0f", rows / durationSeconds) + " γραμμές/δευτ.");
    }
}

File: Benchmarks/CreateCSV.jl
# julia 1.7 secs
# java 2.9 secs
# C++ 1.65 secs

using CSV, Tables, Random, BenchmarkTools

function fast_generate_csv(filename, rows)
    # Ορίζουμε τους τύπους δεδομένων για να βοηθήσουμε τον compiler
    ids = 1:rows
    val1 = rand(Float64, rows)
    val2 = rand(Int32, rows)
    # Δημιουργούμε ένα string buffer για σταθερή κατηγορία (γλιτώνει allocation)
    cat = fill("ABCDE", rows)

    # Χρήση Tables.table για ελάχιστο overhead μνήμης
    tbl = (ID=ids, Value1=val1, Value2=val2, Category=cat)

    println("Έναρξη εγγραφής...")
    # Το CSV.write χρησιμοποιεί αυτόματα threads αν είναι διαθέσιμα
    @time CSV.write(filename, tbl; threaded=true)
end

# Εκτέλεση για 25 εκατομμύρια γραμμές
fast_generate_csv("large_data_julia.csv", 25_000_000)

